### ISSUE TITLE
Tk graphics backend is giving error on Big Sur with DMG installer

### ISSUE THREAD

**Issue by bczottegem** [spyder-ide/spyder#14656]
Changed 'Graphics backend' (Graphics tab) to 'tkinter' and back to 'inline'
On MacOS Big Sur Python 3

### Traceback
```python-traceback
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 236, in button_clicked
    configpage.apply_changes()
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 99, in apply_changes
    self.apply_callback()
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 282, in <lambda>
    self.apply_settings(self.changed_options))
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 369, in apply_plugin_settings
    client_backend_not_inline = [
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 370, in <listcomp>
    client.shellwidget.get_matplotlib_backend() != inline_backend
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/widgets/shell.py", line 341, in get_matplotlib_backend
    return self.call_kernel(
  File "/Applications/Spyder.app/Contents/Resources/lib/python3.9/spyder_kernels/comms/commbase.py", line 550, in __call__
    raise CommError("The comm is not connected.")
spyder_kernels.comms.commbase.CommError: The comm is not connected.
WARNING:traitlets:kernel died: 25.001062870025635
WARNING:traitlets:kernel died: 25.00008487701416
WARNING:traitlets:kernel died: 25.003293991088867
WARNING:traitlets:kernel died: 25.00503396987915
```

## Versions

* Spyder version: 4.2.1 
* Python version: 3.9.1
* Qt version: 5.12.10
* PyQt5 version: 5.12.3
* Operating System: Darwin 20.2.0

### Dependencies

```

# Mandatory:
applaunchservices >=0.1.7       :  0.2.1 (OK)
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  4.0.0 (OK)
cloudpickle >=0.5.0             :  1.6.0 (OK)
diff_match_patch >=20181111     :  20200713 (OK)
intervaltree >=3.0.2            :  3.1.0 (OK)
IPython >=7.6.0                 :  7.19.0 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
keyring >=17.0.0                :  21.5.0 (OK)
nbconvert >=4.0                 :  6.0.7 (OK)
numpydoc >=0.6.0                :  1.1.0 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.7.3 (OK)
pygments >=2.0                  :  2.7.3 (OK)
pylint >=1.0                    :  2.6.0 (OK)
pyls >=0.36.2;<1.0.0            :  0.36.2 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.3.0             :  0.3.0 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  1.0.2 (OK)
qtconsole >=5.0.1               :  5.0.1 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
setuptools >=39.0.0             :  51.0.0 (OK)
sphinx >=0.6.6                  :  3.3.1 (OK)
spyder_kernels >=1.10.1;<1.11.0 :  1.10.1 (OK)
textdistance >=4.2.0            :  4.2.0 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog >=0.10.3               :  1.0.2 (OK)
zmq >=17                        :  20.0.0 (OK)

# Optional:
cython >=0.21                   :  0.29.21 (OK)
matplotlib >=2.0.0              :  3.3.3 (OK)
numpy >=1.7                     :  1.19.4 (OK)
pandas >=1.1.1                  :  None (NOK)
scipy >=0.17.0                  :  1.5.4 (OK)
sympy >=0.7.3                   :  1.7.1 (OK)
```

**Comment by andfoy** [comment]
Hi @bczottegem, thanks for opening this issue. @mrclary, do you know if this error occurs with the new Mac installers

**Comment by mrclary** [comment]
with macOS 11.1 and Spyder 4.2.1 (macOS app build on Catalina) I get the following when switching from "inline" to "Tkinter":

<details><summary>IPython Error</summary>

```
Kernel died, restarting


 
macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm: 57df698261c011eba8ed84383557c522
macOS 11 or later required !
macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm: 53a4fb1461c211eba8ed84383557c522
macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm: 556ee7de61c211eba8ed84383557c522

macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm: 57df698261c011eba8ed84383557c522
macOS 11 or later required !
macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm: 53a4fb1461c211eba8ed84383557c522
macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm: 556ee7de61c211eba8ed84383557c522
macOS 11 or later required !
```
</details>

I do _not_ see any errors switching to "Tkinter" when bootstrapping from conda environment with latest 4.x commit 26cbc8c2b14d00d8639cebb716c2b49d0d8d6819.

The Spyder macOS app artifact from that same commit produces the following on IPython startup

<details><summary>IPython startup</summary>

```
Python 3.9.1 (default, Jan  8 2021, 17:17:17) 
Type "copyright", "credits" or "license" for more information.

IPython 7.19.0 -- An enhanced Interactive Python.

[SpyderKernelApp] WARNING | Could not copy README_STARTUP to startup dir. Source file /Users/rclary/Applications/Spyder.app/Contents/Resources/lib/python39.zip/IPython/core/profile/README_STARTUP does not exist.
[SpyderKernelApp] WARNING | Could not copy README_STARTUP to startup dir. Source file /Users/rclary/Applications/Spyder.app/Contents/Resources/lib/python39.zip/IPython/core/profile/README_STARTUP does not exist.

In [1]:
```
</details>

And then the following on switching from "inline" to "Tkinter"

<details><summary>Tkinter</summary>

```
[SpyderKernelApp] WARNING | Could not copy README_STARTUP to startup dir. Source file /Users/rclary/Applications/Spyder.app/Contents/Resources/lib/python39.zip/IPython/core/profile/README_STARTUP does not exist.
[SpyderKernelApp] WARNING | Could not copy README_STARTUP to startup dir. Source file /Users/rclary/Applications/Spyder.app/Contents/Resources/lib/python39.zip/IPython/core/profile/README_STARTUP does not exist.
ERROR:tornado.application:Exception in callback functools.partial(<function Kernel.enter_eventloop.<locals>.advance_eventloop at 0x11e815e50>)
Traceback (most recent call last):
  File "tornado/ioloop.pyc", line 741, in _run_callback
  File "ipykernel/kernelbase.pyc", line 311, in advance_eventloop
  File "ipykernel/eventloops.pyc", line 220, in loop_tk
  File "tkinter/__init__.pyc", line 2262, in __init__
_tkinter.TclError: Can't find a usable init.tcl in the following directories: 
    /usr/local/Cellar/tcl-tk/8.6.11/lib/tcl8.6 /Users/rclary/Applications/Spyder.app/Contents/lib/tcl8.6 /Users/rclary/Applications/Spyder.app/lib/tcl8.6 /Users/rclary/Applications/Spyder.app/Contents/library /Users/rclary/Applications/Spyder.app/library /Users/rclary/Applications/Spyder.app/tcl8.6.11/library /Users/rclary/Applications/tcl8.6.11/library



This probably means that Tcl wasn't installed properly.

In [1]:
```
</details>

I haven't tried building the macOS application on Big Sur yet. When I do, I'll see if anything here changes.

**Comment by ccordoba12** [comment]
Thanks @mrclary for your help!

> The Spyder macOS app artifact from that same commit produces the following on IPython startup

Perhaps we need to move IPython our of zipped libraries to avoid that. What do you think?

> And then the following on switching from "inline" to "Tkinter"

I'd say the last error shows that we're not packaging Tcl in the app. Are we doing it or not?

**Comment by mrclary** [comment]
> Thanks @mrclary for your help!
> 
> > The Spyder macOS app artifact from that same commit produces the following on IPython startup
> 
> Perhaps we need to move IPython our of zipped libraries to avoid that. What do you think?

That is correct. I had erroneously changed it after the 4.2.1 release. I'll revert that change.

> > And then the following on switching from "inline" to "Tkinter"
> 
> I'd say the last error shows that we're not packaging Tcl in the app. Are we doing it or not?

I don't know what that is, and I've not come across it before...

Nevertheless, I was able to build successfully on Big Sur without requiring any changes to the build process; but I did restore IPython out of the zipped library. The result was the same as for 4.2.1 release: no reference to Tcl or README_STARTUP, just
```
macOS 11 or later required !
[SpyderKernelApp] WARNING | No such comm
```

Interestingly, I don't actually restart the kernel when changing the graphics backend. When switching to Tkinter, I just click apply without exiting the preferences and without applying it to existing kernels.

From the internal console:
<details><summary>Traceback</summary>

```python
Spyder Internal Console

This console is used to report application
internal errors and to inspect Spyder
internals with the following commands:
  spy.app, spy.window, dir(spy)

Please don't use it to run your code

>>> WARNING:traitlets:kernel restarted
WARNING:traitlets:kernel restarted
WARNING:traitlets:kernel restarted
WARNING:traitlets:kernel restarted
WARNING:traitlets:KernelRestarter: restart failed

>>> WARNING:traitlets:kernel died: 25.001260995864868

>>> Traceback (most recent call last):
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 236, in button_clicked
    configpage.apply_changes()
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 99, in apply_changes
    self.apply_callback()
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 282, in <lambda>
    self.apply_settings(self.changed_options))
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 369, in apply_plugin_settings
    client_backend_not_inline = [
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 370, in <listcomp>
    client.shellwidget.get_matplotlib_backend() != inline_backend
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/widgets/shell.py", line 341, in get_matplotlib_backend
    return self.call_kernel(
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder_kernels/comms/commbase.py", line 554, in __call__
    return self._comms_wrapper._get_call_return_value(
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/comms/kernelcomm.py", line 138, in _get_call_return_value
    raise RuntimeError("Kernel is dead")
RuntimeError: Kernel is dead
WARNING:traitlets:kernel died: 25.004613161087036
WARNING:traitlets:kernel died: 25.00042200088501
WARNING:traitlets:kernel restarted
WARNING:traitlets:kernel restarted
WARNING:traitlets:kernel restarted
WARNING:traitlets:kernel restarted
WARNING:traitlets:KernelRestarter: restart failed

>>> WARNING:traitlets:kernel died: 25.004506826400757
Traceback (most recent call last):
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 236, in button_clicked
    configpage.apply_changes()
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 99, in apply_changes
    self.apply_callback()
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 282, in <lambda>
    self.apply_settings(self.changed_options))
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 369, in apply_plugin_settings
    client_backend_not_inline = [
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 370, in <listcomp>
    client.shellwidget.get_matplotlib_backend() != inline_backend
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/widgets/shell.py", line 341, in get_matplotlib_backend
    return self.call_kernel(
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder_kernels/comms/commbase.py", line 550, in __call__
    raise CommError("The comm is not connected.")
spyder_kernels.comms.commbase.CommError: The comm is not connected.
Traceback (most recent call last):
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 227, in accept
    configpage.apply_changes()
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 99, in apply_changes
    self.apply_callback()
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/preferences/configdialog.py", line 282, in <lambda>
    self.apply_settings(self.changed_options))
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 369, in apply_plugin_settings
    client_backend_not_inline = [
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/plugin.py", line 370, in <listcomp>
    client.shellwidget.get_matplotlib_backend() != inline_backend
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder/plugins/ipythonconsole/widgets/shell.py", line 341, in get_matplotlib_backend
    return self.call_kernel(
  File "/Users/rclary/Documents/Python/spyder/installers/macOS/dist/Spyder.app/Contents/Resources/lib/python3.9/spyder_kernels/comms/commbase.py", line 550, in __call__
    raise CommError("The comm is not connected.")
spyder_kernels.comms.commbase.CommError: The comm is not connected.
WARNING:traitlets:kernel died: 25.00447988510132

>>> 
```
<details>

**Comment by mrclary** [comment]
It also seems to only affect the internal environment (Same as Spyder's). When using an external environment, everything seems to behave normally. If anyone has any ideas...

**Comment by ccordoba12** [comment]
@impact27, any ideas about this?

**Comment by mrclary** [comment]
Perhaps some library that is used by or packaged with Tkinter that isn't handled properly? Why would something report "macOS 11 or later required!" when it was run _and built_ on macOS 11?

**Comment by ccordoba12** [comment]
Yeah, that's my feeling too.

**Comment by impact27** [comment]
`spyder_kernels.comms.commbase.CommError: The comm is not connected.` is expected if the kernel crashes. So all errors such as "No such comm" can be ignored. "macOS 11 or later required !" is printed after the kernel crash so it is probably not related either. So the question is: what makes the kernel crash? I think this requires debugging and seeing at what point the kernel dies.

**Comment by mrclary** [comment]
@impact27, do you have access to macOS 11? I can look in the lsp logs, or if you have a better suggestion...

**Comment by mrclary** [comment]
Okay, this is definitely a Tkinter library issue. See [here](https://bugs.python.org/issue42480).
I think I may have found a solution [here](https://stackoverflow.com/questions/60469202/unable-to-install-tkinter-with-pyenv-pythons-on-macos).
Please stand by...

**Comment by mrclary** [comment]
Okay. I can test the `Tkinter` libraries with the following snippet
```python
import tkinter
tkinter.Tk()
```
I get either a gui window (if it works properly); a gui window with deprecation warnings; or our beloved "macOS 11 or later required! Abort trap: 6"
Here are the scenarios:
* Catalina, Homebrew python 3.9.1: works! This is built with Tcl/Tck versions 8.6 (good!)
* Catalina, pyenv python 3.9.0: works! This is built with Tcl/Tck versions 8.5 (bad!). I get the warning:
    ```
    DEPRECATION WARNING: The system version of Tk is deprecated and may be removed in a future release. Please don't rely on it. Set TK_SILENCE_DEPRECATION=1 to suppress this warning.
    ```
* Big Sur, Homebrew python 3.9.1: works! This is built with Tcl/Tck versions 8.6 (good!)
* Big Sur, pyenv python 3.9.0: "macOS 11 or later reqired ! Abort trap: 6". This is built with Tcl/Tck versions 8.5 (bad!)

Now, regarding the macOS app build, I know the following:
* I _was_ able to resolve this issue by installing python in Big Sur via pyenv with the correct `PYTHON_CONFIGURE_OPTS` as mentioned [here](https://stackoverflow.com/questions/60469202/unable-to-install-tkinter-with-pyenv-pythons-on-macos) and building the macOS app on Big Sur (yea!).
* The macOS app built on Catalina in our GitHub actions (macos-latest) uses a Homebrew python built with Tcl/Tck versions 8.5 ! This will certainly cause the issue on Big Sur; but it remains to be tested whether updating the Tcl/Tck versions on Catalina will fix the issue for Big Sur, i.e. we may need both the correct Tcl/Tck versions _and_ to build it on Big Sur...

So, what do we do about it? Homebrew's python _should_ be built with Tcl/Tck versions 8.6. I don't know when Homebrew made this update, but I can only conclude that 'macos-latest' is an outdated image (Docker?). It may be possible to fix this by simply `$ brew uninstall python; brew install python` in our workflow; I'll test this asap. However, this may increase the build time significantly.

When GitHub offers macOS 11 to build on, this may be a moot issue, since undoubtedly they will use an updated Homebrew in any new image they create. However, if we can get the Catalina build to work, we may only need to build one version for both operating systems instead of two (will the Big Sur build app work on Catalina? earlier versions?).

**Comment by mrclary** [comment]
Okay, after all that, it seems that this issue is already resolved.
* I've checked that the latest artifact from #14667 _does_ have Tcl/Tck versions 8.6 and this issue _does not_ manifest on Big Sur.
* I've checked that our latest release 4.2.1 _was_ built with Tcl/Tck versions 8.5 and this issue _does_ manifest on Big Sur.

So, I conclude that some time recently Homebrew updated their Python build to use Tcl/Tck versions 8.6 and macos-latest GitHub action also subsequently updated its image and no action is required on our part.
